Onboarding WPF's Grid control

jmacato opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Ports the Grid control from dotnet/wpf repository. 
## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
`Grid`'s behavior should now be inline with WPF's, hence fixing some long-running issues with our own `Grid` control. 

## Checklist

- [x] Added unit tests (if possible)? 

## Fixed issues
Fixes #2138 (i tested the described scenarios by @donandren but this needs another pair of eyes).

Commit:
Import Grid code from WPF.Core

Commit:
Porting part 1 of n

Commit:
part 2 of n

Commit:
Uncomment Avalonia's Grid and exclude it from build temporarily.

Commit:
Part 3 of n

Gillibald commented on src/Avalonia.Controls/GridWPF.cs:
```diff
@@ -0,0 +1,3854 @@
+        }
+
+        // original implementation, used from 3.0 through 4.6.2
+        private void ResolveStarLegacy(
```
We should probably leave out the old implementation

Commit:
Part 4 of n
Removed more legacy stuff
Added Math Functions needed from WPF's DoubleUtil.

Commit:
Fix misplaced compile command on proj.

Commit:
Part 5 of n

Commit:
Part 6 of n

Commit:
Part 7 of n

Commit:
Part 8 of preliminary error-plugging, Grid running
but somewhat broken.

jmacato commented:
WPF's Grid running on ControlCatalog (still broken though)
![image](https://user-images.githubusercontent.com/16554748/58310845-2f671b00-7e3a-11e9-98bc-4a8f77d3e87c.png)

Commit:
Part 9 of n

Commit:
Temporarily hook row and col def change to ExtData.

jmacato commented:
Getting there! (calendar broke for some reason)
![image](https://user-images.githubusercontent.com/16554748/58384568-c0461e00-8015-11e9-8700-e250fa81f766.png)

Commit:
Part 10 of n

Commit:
Fix rowcol definitions not sync'ing with DefinitionsU/V, hence fixing bugs for the Calendar control.

Commit:
Cleaning up redundant/unreferenced code.

Gillibald commented:
I don't think we should add class handlers for column, row attached to each visual. Instead do this when the property is attached to a visual.

jmacato commented:
@Gillibald Yes, thats the right way. There's still some temporary hacks and that is one of 'em so thanks for the reminder :smile:

Commit:
Fix DPI fallback value on LayoutRounding;
Rename `rowColDefsEmpty` to `IsTrivialGrid`;

Commit:
Replace per-property class handlers with `AffectsParentMeasure<>`; Fix `GridLinesRenderer`.

jmacato commented:
`ShowGridLines` now working
![image](https://user-images.githubusercontent.com/16554748/58393786-ddf5a080-8072-11e9-92de-d513853daf02.png)

Commit:
Rename private fields, remove practically unused fields.

Commit:
Merge branch 'master' into wpfgridimport

Commit:
Merge branch 'wpfgridimport' of github.com:AvaloniaUI/Avalonia into wpfgridimport

Commit:
Downsize DefinitionBase, removing most of shared size scope code.

Commit:
Calculate ActualWidth/Height for Row/ColDefinitions so that GridSplitter can work.

Commit:
Reset MinSize on DefinitionsLayout validation.

Commit:
Cleanup MinSize.

Commit:
Some refactoring...

Commit:
Disable SharedSize tests for now.

Commit:
Disable Shared Size Tests pt. 2

Commit:
Remove old Grid classes.

Commit:
Revert `DrawGridLines` to FinalOffsets

Commit:
Organize Grid into its own folder and split up its internal classes.

Commit:
Re-run CI

jmacato commented:
Almost all features but `SharedSizeScopes` is working now :smile:

Gillibald commented on src/Avalonia.Controls/Grid/GridLinesRenderer.cs:
```diff
@@ -0,0 +1,95 @@
+        static GridLinesRenderer()
+        {
+            var oddDashArray = new List<double>();
+            oddDashArray.Add(_dashLength);
```
even and odd are the same

jmacato replied:
Nice catch! Double-checked the original source, i've mangled it while copypasting it seems

Commit:
Fix GridLineRenderer's DashStyles.

Commit:
More reorganizing + refactoring...

Commit:
Revert `ShowGridLines` default value.

Commit:
Implement and make SharedSizeScopes work :)

jmacato commented:
Grid is now fully-functional :label: 
![image](https://user-images.githubusercontent.com/16554748/58463604-bf70d180-8166-11e9-83fb-7323239feb44.png)

x2bool commented:
Great job! I hope all of those small bugs of the old Grid implementation will be gone now.

Commit:
Inherit SharedSize scopes to child grids.

Commit:
Add SharedSizeScope tests from #1945
Fix IsSharedSizeScope/PrivateSharedSizeScope so that it can be
set on non-`Grid` controls.

jmacato commented on src/Avalonia.Controls/Grid/DefinitionBase.cs:
```diff
@@ -0,0 +1,342 @@
+            _parentIndex = index;
+        }
+
+        internal void UpdateSharedScope()
```
@grokys This function is being called all over the place, basically i need both PrivateSharedSizeScope and SharedSizeGroup not null when this runs so i resorted in placing calls on this method on both Measure/Arrange which is not ideal...

jmacato commented on src/Avalonia.Controls/Grid/Grid.cs:
```diff
@@ -0,0 +1,2373 @@
+                return _columnDefinitions;
+            }
+            set
+            {
```
@grokys This and `RowDefinition`'s setter are horrible but i ran out of ideas on how to properly mirror those two on their array counterparts...

jmacato commented on tests/Avalonia.Controls.UnitTests/GridTests.cs:
```diff
@@ -180,5 +183,267 @@ public void Changing_Child_Column_Invalidates_Measure()
+
+        [Fact]
+        public void Size_Is_Propagated_Between_Rows_And_Columns()
+        {
```
@grokys This baffles me sir as i really have no idea why this doesn't work at all

Commit:
Restore unit tests part 2

Commit:
Revert back to WPF and start from scratch, using thepower of Hindsight~

Commit:
Trim header comments

Commit:
Remove pragma and GridParanoia regions and references to that code.

Commit:
Simplify usings.

Commit:
Remove IAddChild and its interface implementations

Commit:
Replace header legalese with the appropiate one.

Commit:
Fix usings and namespace on DefBase

Commit:
Remove regions and remaining Pragmas + Remove Localization attribute in DefBase.

jmacato commented:
@AvaloniaUI/core please don't update this branch to master until this second phase is done.

Commit:
Replace some Types from WPF to Avalonia equivalents.

Commit:
Remove AttachedPropertyBrowsableForChildren attribute

Commit:
Remove DesignerSerializationVisibility attribute

Commit:
Replace invariant.assert with Contract.Requires.

Commit:
Replace WPF types with equivalent Avalonia ones part 2.

Commit:
Add missing imports.

Commit:
Remove EditorBrowsable attribute

Commit:
Convert DependencyProperties to Avalonia's.

Commit:
Restore deleted GridLength.cs

Commit:
Fix GridLinesRenderer

Commit:
Comment-out code that is not going to be used.

Commit:
Disable legacy GridLength.Star algorithms.
Turn DefBase[] into IReadOnlyList.
Fix DefBase, turn it into AvaloniaObject, change WPF types to Avalonia's and add a Parent prop.

Commit:
Refactor child references.

Commit:
Add WPF's UIElement.RoundLayoutValue to MathUtilities.

Commit:
Fix Grid dpi and roundedlayout stuff + add GridLinesRenderer as Grid's visual child.

Commit:
Fix definitionbase attached properties.

Commit:
Fix OnCellAttachedPropertyChanged

Commit:
Restore removed GridSplitter code.

Commit:
Make ValueCache as abstract props.
Add setters to Row/ColumnDefinitions.

Commit:
Trigger OnEnter/EnterParentTree on DefBase.

Commit:
Invalidate parent grid when property changes on DefBase.
Link ActualWidth/Height setters to their parent GetFinalColumnDefinitionWidth/GetFinalRowDefinitionWidth

Commit:
Add inheritance parent to DefBase.

Commit:
Fix shared size scope handlers. 5/9 unit tests on SharedSizeScopeTests now works.

Commit:
Pass all of GridTests.

Commit:
Added tests from WpfGridTests, Special thanks to @wieslawsoltes 🎉

Commit:
Add some tests

Commit:
Update GridMultiTests

Commit:
Fixes defBase behavior when removed/added from collection.

Commit:
Uncomment a disabled unit-test.

Commit:
Make a new base class for both Row/ColumnDefinitions.

Commit:
Fix unit tests and transfer back to GridTests.cs

grokys reviewed (approved):
Looks good to me, it's pretty hard to review as it's a code dump from WPF, but gets a 👍 from me!

grokys commented:
Ah, seems there's a failing test?

grokys committed:
Merge branch 'master' into wpfgridimport

grokys committed:
Merge branch 'master' into wpfgridimport

Commit:
Dont flag the Row/ColDefinition list dirty when the child Row/ColDefinition's child property is changed.
